realm/realm-library/src/androidTest/kotlin/io/realm/EmbeddedObjectsTest.kt

```diff
@@ -527,21 +528,155 @@ class EmbeddedObjectsTest {
+                        assertEquals(list[0].child!!.childId, it[0]!!.childId)
+                        assertEquals(list[1].child!!.childId, it[1]!!.childId)
+                        assertEquals(list[2].child!!.childId, it[2]!!.childId)
+                        assertEquals(list[0]._id, it[0]!!.parent._id)
+                        assertEquals(list[1]._id, it[1]!!.parent._id)
+                        assertEquals(list[2]._id, it[2]!!.parent._id)
+                    }
+        }
+    }
+
+    @Test
+    fun insert_listWithEmbeddedObjects_duplicatePrimaryKeyThrows() {
+        realm.executeTransaction { realm ->
+            val list = listOf<EmbeddedSimpleParent>()
+                    .plus(EmbeddedSimpleParent("parent1").apply { child = EmbeddedSimpleChild("child1") })
+                    .plus(EmbeddedSimpleParent("parent2").apply { child = EmbeddedSimpleChild("child2") })
+                    .plus(EmbeddedSimpleParent("parent3").apply { child = EmbeddedSimpleChild("child3") })
+                    .also { realm.insert(it) }
+
+            assertFailsWith<RealmPrimaryKeyConstraintException> {
+                realm.insert(list)
+            }
+        }
+    }
+
+    @Test
+    fun insert_listWithEmbeddedObjects_insertingChildrenDirectlyThrows() {
+        realm.executeTransaction { realm ->
+            listOf<EmbeddedSimpleChild>()
+                    .plus(EmbeddedSimpleChild("child1"))
+                    .plus(EmbeddedSimpleChild("child2"))
+                    .plus(EmbeddedSimpleChild("child3"))
+                    .also {
+                        assertFailsWith<IllegalArgumentException> {
+                            realm.insert(it)
+                        }
+                    }
+        }
     }
 
     @Test
-    @Ignore("Add in another PR")
     fun insertOrUpdate_listWithEmbeddedObjects() {
-        TODO()
+        realm.executeTransaction { realm ->
+            val list = listOf<EmbeddedSimpleParent>()
+                    .plus(EmbeddedSimpleParent("parent1").apply { child = EmbeddedSimpleChild("child1") })
+                    .plus(EmbeddedSimpleParent("parent2").apply { child = EmbeddedSimpleChild("child2") })
+                    .plus(EmbeddedSimpleParent("parent3").apply { child = EmbeddedSimpleChild("child3") })
+                    .also { realm.insert(it) }
+            val managedChildren = realm.where<EmbeddedSimpleChild>().findAll()
+
+            val newList = listOf<EmbeddedSimpleParent>()
+                    .plus(EmbeddedSimpleParent("parent1").apply { child = EmbeddedSimpleChild("newChild1") })
+                    .plus(EmbeddedSimpleParent("parent2").apply { child = EmbeddedSimpleChild("newChild2") })
+                    .plus(EmbeddedSimpleParent("parent3").apply { child = EmbeddedSimpleChild("newChild3") })
+                    .also { realm.insertOrUpdate(it) }
+
+            assertNull(realm.where<EmbeddedSimpleChild>().equalTo("childId", list[0].child!!.childId).findFirst())
```

Comment:
Sometimes it is a bit dangerous to reference the expected value from a complex expression. Typing it explicitly or referencing a local string variable makes it more clear.

Reply:
I give that here it is probably easier to make an error by putting the values manually instead of referencing it.